plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -1879,6 +1879,17 @@ producer_plugin_impl::start_block_result producer_plugin_impl::start_block() {
       _pending_block_mode = pending_block_mode::speculating;
    }
 
+   if (in_speculating_mode()) {
+      static fc::time_point last_start_block_time = fc::time_point::maximum(); // always start with speculative block
+      // Determine if we are syncing: if we have recently started an old block then assume we are syncing
+      if (last_start_block_time < now + fc::microseconds(config::block_interval_us)) {
+         auto head_block_age = now - chain.head().block_time();
+         if (head_block_age > fc::seconds(5))
+            return start_block_result::waiting_for_block; // if syncing no need to create a block just to immediately abort it
+      }
+      last_start_block_time = now;
+   }
+
```

Comment:
moving this block before means it will not be executed when we set `_pending_block_mode = pending_block_mode::speculating;` below. Is this what we want?

Reply:
Yes, that is exactly why it was moved up.